examples/medplum-demo-bots/src/deduplication/merge-matching-patients.ts

```diff
@@ -0,0 +1,199 @@
+  await updateClinicalReferences(medplum, mergedPatients.src, mergedPatients.target, 'DiagnosticReport');
+  await updateClinicalReferences(medplum, mergedPatients.src, mergedPatients.target, 'MedicationRequest');
+  await updateClinicalReferences(medplum, mergedPatients.src, mergedPatients.target, 'Encounter');
+
+  // We might delete the source patient record if we don't want to continue to have a duplicate of an existing patient
+  // despite the fact that it is an inactive record.
+  const deleteSource = responses['deleteSource']?.valueBoolean;
+  if (deleteSource === true) {
+    await medplum.deleteResource('Patient', mergedPatients.src.id as string);
+  } else {
+    // If we don't delete the source patient record, we need to update it to be inactive.
+    await medplum.updateResource<Patient>(mergedPatients.src);
+  }
+
+  // Update the target patient record with the merged data, and have it as the master record.
+  await medplum.updateResource<Patient>(mergedPatients.target);
+}
+
+interface MergedPatients {
+  readonly src: Patient;
+  readonly target: Patient;
+}
+
+type HasSubject = {
+  subject?: Reference<Patient>;
+};
+
+type ResourceWithSubject = Resource & HasSubject;
+
+/**
+ * Links two patient records indicating one replaces the other.
+ *
+ * @param src - The source patient record which is being replaced.
+ * @param target - The target patient record which will replace the source.
+ * @returns - Object containing updated source and target patient records with their links.
+ */
+export function linkPatientRecords(src: Patient, target: Patient): MergedPatients {
+  const targetCopy = deepClone(target);
+  const targetLinks = targetCopy.link ?? [];
+  targetLinks.push({ other: createReference(src), type: 'replaces' });
+
+  const srcCopy = deepClone(src);
+  const srcLinks = srcCopy.link ?? [];
+  srcLinks.push({ other: createReference(target), type: 'replaced-by' });
+  return { src: { ...srcCopy, link: srcLinks, active: false }, target: { ...targetCopy, link: targetLinks } };
+}
+
+/**
+ * Merges contact information (identifiers) of two patient records, where the source patient record will be marked as an old record.
+ * The target patient record will be overwritten with the merged data and will be the master record.
+ *
+ * @param src - The source patient record.
+ * @param target - The target patient record.
+ * @param fields - Optional additional fields to be merged.
+ * @returns - Object containing the original source and the merged target patient records.
+ */
+export function mergePatientRecords(src: Patient, target: Patient, fields?: Partial<Patient>): MergedPatients {
+  const srcIdentifiers = src.identifier ?? [];
+  const mergedIdentifiers = srcIdentifiers.map((identifier) => ({
+    ...identifier,
+    use: 'old' as Identifier['use'],
+  }));
+  const targetCopy = deepClone(target);
+  targetCopy.identifier = [...(targetCopy.identifier ?? []), ...mergedIdentifiers];
```

Comment:
This might be confusing if both new and old records share an identifier: one could have e.g.:
```json
"identifier": [
  {
    "system": "http://example.com/id",
    "value": "foo",
    "use": "official"
  },
  {
    "system": "http://example.com/id",
    "value": "foo",
    "use": "old"
  }
]
```

If the identifier is the same in the target resource, maybe the merge should just keep that element and avoid adding a duplicate.